src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -300,8 +311,19 @@ private ReadOnlyMemory<byte> ReadChunkFromConnectionBuffer(int maxBytesToRead, C
                                     _state = ParsingState.Done;
                                     _connection.CompleteResponse();
                                     _connection = null;
+
+                                    // Reached EOF.
+                                    _response.TrailingHeaders = _trailers ??
```

Comment:
It seems weird that we actually create the TrailingHeaders collection here. We don't do this for regular headers. We just let the HttpResponseMessage object create this on demand for us. Seems like we should follow the same pattern here.

Reply:
> We just let the HttpResponseMessage object create this on demand for us. Seems like we should follow the same pattern here.

I can do that - the tricky part is how to let **HttpResponseMessage** to create this collection **on demand** to avoid allocation, since we need to read/buffer till the stream EOF to know if we need to create a new `TrailingHeaders` collection - and we need to return the `HttpResponseMessage` once the `Header` info is ready. (depends on the CompletionOption.Header)

If we can allocate a new collection for each `HttpResponseMessage` will simplify things.